---
title: "03.00-D-sRNAseq-gene-expression-DESeq2"
author: "Sam White"
date: "2025-01-31"
output: 
  bookdown::html_document2:
    theme: cosmo
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    code_download: true
  github_document:
    toc: true
    number_sections: true
  html_document:
    theme: cosmo
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    code_download: true
bibliography: references.bib
---

# Background

This will run [DEseq2](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) [@love2014] to determine if any of the miRNAs identified in [02.00-ShortStack-31bp-fastp-merged.md](https://github.com/RobertsLab/project-clam-oa/blob/2616d45a5a0bbef23081c5824bfe5bad6b89d227/code/02.00-ShortStack-31bp-fastp-merged.md) analysis are differentially expressed between control/treatment.

This was initially run with a log2 fold change threshold set to 1 (which is equivalent to a 2-fold change in expression), but that returned 0 differentially expressed miRNAs. As such, this was run again with the log2 fold change threshold set to 0.

::: callout-note
ShortStack identified 37 miRNAs.
:::

## Inputs

-   [`Counts.txt`](https://github.com/RobertsLab/project-clam-oa/blob/d380ed9a88c874db5434c9ec74f6ab1de042bad5/output/02.00-ShortStack-31bp-fastp-merged/ShortStack_out/Counts.txt): ShortStack counts matrix. Includes all clusters, including those that were *not* categorized as miRNAs.

-   [`DESeq2-coldata.tab`](https://github.com/RobertsLab/project-clam-oa/blob/d380ed9a88c874db5434c9ec74f6ab1de042bad5/output/03.00-D-sRNAseq-gene-expression-DESeq2/DESeq2-coldata.tab): Two column table with sample ID and treatment. This file is also an output from this notebook.

-   [ManilaOA2023_shortRNASeq_Meta.csv](https://github.com/RobertsLab/project-clam-oa/blob/d380ed9a88c874db5434c9ec74f6ab1de042bad5/data/ManilaOA2023_shortRNASeq_Meta.csv): Metadata file for this sRNA-seq data.

## Outputs

-   [`DESeq2-coldata.tab`](https://github.com/RobertsLab/project-clam-oa/blob/d380ed9a88c874db5434c9ec74f6ab1de042bad5/output/03.00-D-sRNAseq-gene-expression-DESeq2/DESeq2-coldata.tab): Two column table with sample ID and treatment. Needed as input to DEseq2.

-   [`deseq2.table.csv`](https://github.com/RobertsLab/project-clam-oa/blob/d380ed9a88c874db5434c9ec74f6ab1de042bad5/output/03.00-D-sRNAseq-gene-expression-DESeq2/deseq2.table.csv): DEseq2 output table containing mean expression, fold change in expression, and adjusted p-values for all input samples.

-   [deseq2.fdr-0.05.lfc-0.table.csv](https://github.com/RobertsLab/project-clam-oa/blob/3b23f7cfcdf9cef627556090518d60d685d881bd/output/03.00-D-sRNAseq-gene-expression-DESeq2/deseq2.fdr-0.05.lfc-0.table.csv): Subset of DEseq2 output table of all miRNA clusters ("genes") having adjusted p-values (fdr) \<= 0.05 with log2 fold change cutoff set to 0.

```{r setup, include=FALSE}
library("RColorBrewer")
library("DESeq2")
library("ggplot2")
library("knitr")
library("pheatmap")
library("tidyverse")
knitr::opts_chunk$set(
  echo = TRUE,         # Display code chunks
  eval = FALSE,        # Evaluate code chunks
  warning = FALSE,     # Hide warnings
  message = FALSE,     # Hide messages
  comment = ""         # Prevents appending '##' to beginning of lines in code output
)
```

# Set R variables

```{r set-variables, eval=TRUE}
# Define the output directory path
output_dir <- "../output/03.00-D-sRNAseq-gene-expression-DESeq2/"

# Set desired false discovery rate threshold (i.e. adjusted p-value, padj)
fdr <- 0.05

# Set log2 fold change threshold (a value of '1' is equal to a fold change of '2')
log2fc <- 0
```

# Load count data

Load in the sRNA count matrix generated using ShortStack 4.1.1. Keep in mind this data includes counts of all sRNAs, not just miRNAs.

Counts generated in `02.00-ShortStack-31bp-fastp-merged`.

## Select only miRNAs IDd by ShortStack

```{r load-sRNA-counts, eval=TRUE}
# Read in sRNA counts data
srna_seq_counts_shortstack <- read_delim("../output/02.00-ShortStack-31bp-fastp-merged/ShortStack_out/Counts.txt", delim="\t")

srna_seq_counts_shortstack <- srna_seq_counts_shortstack %>% filter(MIRNA == "Y")

str(srna_seq_counts_shortstack)

```

# Create DESeq2 Column Data

## Read in metadata CSV

```{r read-metadata, eval=TRUE}
# Load metadata
metadata <- read.csv("../data/ManilaOA2023_shortRNASeq_Meta.csv", header = TRUE)

str(metadata)
```

## Extract sample names

```{r extract-sample-names, eval=TRUE}
sample_names <- colnames(srna_seq_counts_shortstack) %>%
  str_subset("^\\d+-") %>%
  str_extract("^\\d+")

str(sample_names)
```

## Select sample name and treatment

```{r filter-metadata, eval=TRUE}
sample_treatment_df <- metadata %>%
  select(ID_simple, treatment)

# Set sample names as rownames
rownames(sample_treatment_df) <- sample_treatment_df$ID_simple

sample_treatment_df$ID_simple <- NULL

str(sample_treatment_df)
```

## Write DEseq coldata to file

```{r write-deseq-coldata, eval=TRUE}
write.table(
  sample_treatment_df,
  file = "../output/03.00-D-sRNAseq-gene-expression-DESeq2/DESeq2-coldata.tab",
  sep = "\t",
  quote = FALSE,
  col.names = NA
)

```

# Count data munging

## Fix col names and convert to matrix

```{r sRNA-count-data-munging, eval=TRUE}
coldata <- sample_treatment_df

# Remove excess portions of sample column names to just "sample###"
colnames(srna_seq_counts_shortstack) <- sub("-fastp-adapters-polyG-31bp-merged_condensed", "", colnames(srna_seq_counts_shortstack))

# Keep just the counts and cluster names as matrix
srna_seq_counts_matrix <- as.matrix(srna_seq_counts_shortstack %>% select(-Coords, -MIRNA) %>% column_to_rownames(var = "Name"))

str(srna_seq_counts_matrix)
```

## Take only samples present in coldata

```{r extract-samples, eval=TRUE}

common_cols <- intersect(colnames(srna_seq_counts_matrix),
                         rownames(sample_treatment_df))

srna_seq_counts_matrix <- srna_seq_counts_matrix[, common_cols]

str(srna_seq_counts_matrix)
```

## Reorder matrix cols to match coldata

```{r reorder-matrix-cols, eval=TRUE}
ord <- match(rownames(sample_treatment_df), colnames(srna_seq_counts_matrix))

srna_seq_counts_matrix_sorted <- srna_seq_counts_matrix[, ord]

str(srna_seq_counts_matrix_sorted)
```

## Verify rownames match

```{r check-rownames, eval=TRUE}
all(rownames(coldata) == colnames(srna_seq_counts_matrix_sorted))
```

# Create DESeq2 data set

## Initialize DEseq2 data set

```{r create-deseq2-data-set, eval=TRUE}
dds <- DESeqDataSetFromMatrix(countData = srna_seq_counts_matrix_sorted,
                              colData = coldata,
                              design = ~ treatment)
dds

```

## Add cluster column as "gene" feature

```{r add-cluaster-as-gene-feature, eval=TRUE}
featureData <- data.frame(gene=rownames(srna_seq_counts_matrix_sorted))
mcols(dds) <- DataFrame(mcols(dds), featureData)
mcols(dds)
```

## Set factor levels

```{r set-factor-levels, eval=TRUE}
dds$treatment <- factor(dds$treatment, levels = c("control", "treatment"))
```

# DESeq analysis

## DEseq

```{r deseq, eval=TRUE}
dds <- DESeq(dds)
```

## DEseq Results

```{r assign-results, eval=TRUE}

res <- results(dds, alpha = fdr, lfcThreshold = log2fc)

res

summary(res)

table(res$padj < fdr)
```

## Write DDS results tables to CSVs

```{r write-dds-results-table, eval=TRUE}
write.csv(res, file = paste0(output_dir, "deseq2", ".table.csv"), row.names = TRUE, quote = FALSE)

# Subset based on adjusted p-value
resSig <- subset(res, padj < fdr)

resSig
write.csv(resSig, file = paste0(output_dir, "deseq2",".fdr-", fdr, ".lfc-", log2fc, ".table.csv"), row.names = TRUE, quote = FALSE)
```

## Identify Differentially Expressed miRNAs
```{r differentially-expressed-RNAs, engine='bash', eval=TRUE}


# Extract the first column values (excluding the header)
clusters=$(cut -d',' -f1 "../output/03.00-D-sRNAseq-gene-expression-DESeq2/deseq2.fdr-0.05.lfc-0.table.csv" | tail -n +2)

# Loop through each cluster and search in the Results.txt file
for cluster in $clusters; 
do
    grep "$cluster" "../output/02.00-ShortStack-31bp-fastp-merged/ShortStack_out/Results.txt"
done

echo ""
echo "--------------------------------------------------------------"
echo ""

for cluster in $clusters; 
do
    grep "$cluster" "../output/02.00-ShortStack-31bp-fastp-merged/ShortStack_out/Results.txt" \
    | awk '{print $21}'
done
```

# Variance stabilizing transformations (VST)

-   Here we transform counts using a variance stabilizing transformation (VST), since we have many samples.

```{r VST, eval=TRUE}
vsd <- varianceStabilizingTransformation(dds, blind=FALSE)
```

# Plotting

## Sample distances

```{r plot-sample-distances}
sampleDists <- dist(t(assay(vsd)))

sampleDistMatrix <- as.matrix( sampleDists )

rownames(sampleDistMatrix) <- paste( vsd$colony.id, vsd$time.point, sep = " - " )

colnames(sampleDistMatrix) <- NULL

colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)

pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
```

## PCA

Visualize sample clustering via PCA (after transformation)

```{r pca, eval=TRUE}
# PCA with points color coded by time point 
plotPCA(vsd, intgroup = c("treatment"))
```

## Heatmap of all 37 ShortStack miRNAs

```{r heatmap, eval=TRUE}
top40_counts <- order(rowMeans(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:37]

annotation = colData(dds) %>% as.data.frame() %>% select(treatment)

pheatmap(assay(vsd)[top40_counts,], 
         cluster_rows=FALSE, 
         show_rownames=TRUE,
         cluster_cols=TRUE, 
         annotation_col = annotation)

```

# REFERENCES
